### PR TITLE
[Fix] Fix bugs in DynamicScatter op

### PR DESCRIPTION
**Pull request by wHao-Wu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

The unittest of mmdet3d fails after importing the DynamicScatter op from mmcv. I found that there exists an inconsistency of DynamicScatter op between mmdet3d and mmcv, which causes the failure of mmdet3d.

## Modification

Recover the `if` statement.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by ZwwWayne** [comment]
Invite @zhanggefan to take a look at this PR

File: mmcv/ops/csrc/pytorch/cuda/scatter_points_cuda.cu [hunk 1]
```diff
@@ -26,10 +26,15 @@ std::vector<at::Tensor> DynamicPointToVoxelForwardCUDAKernelLauncher(
   std::tie(out_coors, coors_map, reduce_count) =
       at::unique_dim(coors_clean, 0, true, true, true);
 
-  // the first element of out_coors is always (-1,-1,-1) and should be removed
-  out_coors = out_coors.slice(0, 1);
-  reduce_count = reduce_count.slice(0, 1).to(torch::kInt32);
-  coors_map = coors_map.to(torch::kInt32) - 1;
+  if (out_coors.index({0, 0}).lt(0).item<bool>()) {
```
Comment (ZwwWayne) [inline mmcv/ops/csrc/pytorch/cuda/scatter_points_cuda.cu:29]: is there any other cases that need this judgment?

Comment (zhanggefan) [reply mmcv/ops/csrc/pytorch/cuda/scatter_points_cuda.cu:29]: This judgment is necessary if the input may contain out-of-range `coors` (full with -1). If we could guarantee that the input coors are all valid, then we may safely remove this branch.

File: tests/test_ops/test_scatter_points.py [hunk 2]
```diff
@@ -11,7 +11,7 @@ def test_dynamic_scatter():
     feats = torch.rand(
         size=(200000, 3), dtype=torch.float32, device='cuda') * 100 - 50
     coors = torch.randint(
-        low=-1, high=20, size=(200000, 3), dtype=torch.int32, device='cuda')
+        low=0, high=20, size=(200000, 3), dtype=torch.int32, device='cuda')
```
Comment (zhanggefan) [inline tests/test_ops/test_scatter_points.py:15]: why should we change this? I personally think the original test that simulates input with invalid coors (have value -1) is actually beneficial. We actually could not guarantee that the output coors of dynamic voxelization are all valid according to the logics here:
https://github.com/open-mmlab/mmcv/blob/a0127e1712cd58b22736d20ed4e5531e1d277f5e/mmcv/ops/csrc/common/cuda/voxelization_cuda_kernel.cuh#L22-L49

File: mmcv/ops/csrc/pytorch/cuda/scatter_points_cuda.cu (hunk 1 above)
Comment (zhanggefan) [reply mmcv/ops/csrc/pytorch/cuda/scatter_points_cuda.cu:29]: I just check the code of dynamic voxelization and found that we could not rule out the possibility that the input contains out-of-range coors.

Comment (ZwwWayne) [reply mmcv/ops/csrc/pytorch/cuda/scatter_points_cuda.cu:29]: sure, then we can suggest @wHao-Wu to leave a comment here as a reminder.

File: tests/test_ops/test_scatter_points.py (hunk 2 above)
Comment (wHao-Wu) [reply tests/test_ops/test_scatter_points.py:15]: Sorry for the mistake. I found that some of the unittest in mmdet3d has been deleted here. The op cannot pass the unittest after the recovery of the unittest before this PR. Then I need to recover the unittest to implement the strict testing for the DynamicScatter op.

File: mmcv/ops/csrc/pytorch/cuda/scatter_points_cuda.cu (hunk 1 above)
Comment (wHao-Wu) [reply mmcv/ops/csrc/pytorch/cuda/scatter_points_cuda.cu:29]: OK. I will leave a comment here. Thanks for your help.

File: tests/test_ops/test_scatter_points.py (hunk 2 above)
Comment (zhanggefan) [reply tests/test_ops/test_scatter_points.py]: LGTM

**Review by zhanggefan — APPROVED** [review]

File: mmcv/ops/csrc/pytorch/cuda/scatter_points_cuda.cu [hunk 3]
```diff
@@ -26,10 +26,16 @@ std::vector<at::Tensor> DynamicPointToVoxelForwardCUDAKernelLauncher(
   std::tie(out_coors, coors_map, reduce_count) =
       at::unique_dim(coors_clean, 0, true, true, true);
 
-  // the first element of out_coors is always (-1,-1,-1) and should be removed
-  out_coors = out_coors.slice(0, 1);
-  reduce_count = reduce_count.slice(0, 1).to(torch::kInt32);
-  coors_map = coors_map.to(torch::kInt32) - 1;
+  // deal with the condition when the input may contain out-of-range coors (full with -1).
+  if (out_coors.index({0, 0}).lt(0).item<bool>()) {
+    // the first element of out_coors (-1,-1,-1) and should be removed
```
Comment (ZwwWayne) [inline mmcv/ops/csrc/pytorch/cuda/scatter_points_cuda.cu:31]: rephrase the comment

**Review by ZwwWayne — APPROVED** [review]

**Review by grimoire — APPROVED** [review]

**Pull request merged**